pkg/storage/iterators.go

```diff
@@ -411,3 +412,121 @@ func NewConditionsFilteredTupleKeyIterator(iter TupleKeyIterator, filter TupleKe
 func IterIsDoneOrCancelled(err error) bool {
 	return errors.Is(err, ErrIteratorDone) || errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded)
 }
+
+type OrderedCombinedIterator struct {
+	mu      *sync.Mutex
+	once    *sync.Once
+	mapper  TupleMapper
+	pending []TupleIterator // GUARDED_BY(mu)
+}
+
+var _ TupleIterator = (*OrderedCombinedIterator)(nil)
+
+type TupleMapper func(t *openfgav1.Tuple) string
+
+func UserMapper() TupleMapper {
+	return func(t *openfgav1.Tuple) string {
+		return t.GetKey().GetUser()
+	}
+}
+
+func ObjectMapper() TupleMapper {
+	return func(t *openfgav1.Tuple) string {
+		return t.GetKey().GetObject()
+	}
+}
+
+// NewOrderedCombinedIterator is a thread-safe iterator that combines a list of iterators into a single ordered iterator.
+// All the input iterators must be individually ordered already according to mapper.
+// Iterators can yield the same value (as defined by mapper) multiple times, but it will only be returned once.
+func NewOrderedCombinedIterator(mapper TupleMapper, sortedIters ...TupleIterator) *OrderedCombinedIterator {
+	pending := make([]TupleIterator, 0, len(sortedIters))
+	for _, sortedIter := range sortedIters {
+		if sortedIter != nil {
```

Comment:
In what case we would pass `nil` iterators? I believe I've seen this elsewhere, doesn't need to be handled in this PR.

Reply:
In no case. I copied from the other iterators

Reply:
Can we clean that up in a subsequent PR across the others 🤔